assignment-client/src/assets/AssetServer.cpp

```diff
@@ -0,0 +1,279 @@
+                    replyPacketList->write(data, chunkSize);
+                }
+                qDebug() << "Done reading";
+            }
+            file.close();
+        } else {
+            qDebug() << "Asset not found";
+            writeError(replyPacketList.get(), AssetServerError::ASSET_NOT_FOUND);
+        }
+    }
+
+    qDebug() << "Sending asset";
+    auto nodeList = DependencyManager::get<NodeList>();
+    nodeList->sendPacketList(std::move(replyPacketList), *_sendToNode);
+}
+
+AssetServer::AssetServer(NLPacket& packet) :
+        ThreadedAssignment(packet),
+        _taskPool(this) {
+
+    // Most of the work will be I/O bound, reading from disk and constructing packet objects,
+    // so the ideal is greater than the number of cores on the system.
+    _taskPool.setMaxThreadCount(20);
+
+    auto& packetReceiver = DependencyManager::get<NodeList>()->getPacketReceiver();
+    packetReceiver.registerListener(PacketType::AssetGet, this, "handleAssetGet");
+    packetReceiver.registerListener(PacketType::AssetGetInfo, this, "handleAssetGetInfo");
+    packetReceiver.registerMessageListener(PacketType::AssetUpload, this, "handleAssetUpload");
+}
+
+AssetServer::~AssetServer() {
+}
+
+void AssetServer::run() {
+    ThreadedAssignment::commonInit(ASSET_SERVER_LOGGING_TARGET_NAME, NodeType::AssetServer);
+
+    auto nodeList = DependencyManager::get<NodeList>();
+    nodeList->addNodeTypeToInterestSet(NodeType::Agent);
+
+    _resourcesDirectory = QDir(QCoreApplication::applicationDirPath()).filePath("resources/assets");
+    if (!_resourcesDirectory.exists()) {
+        qDebug() << "Creating resources directory";
+        _resourcesDirectory.mkpath(".");
+    }
+    qDebug() << "Serving files from: " << _resourcesDirectory.path();
+
+    // Scan for new files
+    qDebug() << "Looking for new files in asset directory";
+    auto files = _resourcesDirectory.entryInfoList(QDir::Files);
+    QRegExp filenameRegex { "^[a-f0-9]{32}(\\..+)?$" };
+    for (auto fileInfo : files) {
+        auto filename = fileInfo.fileName();
+        if (!filenameRegex.exactMatch(filename)) {
+            qDebug() << "Found file: " << filename;
+            if (!fileInfo.isReadable()) {
+                qDebug() << "\tCan't open file for reading: " << filename;
+                continue;
+            }
+
+            // Read file
+            QFile file { fileInfo.absoluteFilePath() };
+            file.open(QFile::ReadOnly);
+            QByteArray data = file.readAll();
+
+            auto hash = hashData(data);
+
+            qDebug() << "\tMoving " << filename << " to " << hash;
+
+            file.rename(_resourcesDirectory.absoluteFilePath(hash));
+        }
+    }
+
+    while (!_isFinished) {
+        // since we're a while loop we need to help Qt's event processing
+        QCoreApplication::processEvents();
+        // QCoreApplication::processEvents(QEventLoop::WaitForMoreEvents);
+    }
```

Comment:
If you just let this `run` method return it should use the `QThread` event loop to start processing events